transcom/mymove
# [MB-2978] Flash Message reducer & component

## Description

This PR adds a new `flash` reducer that can be used for setting (& clearing) flash messages that will be rendered on the Home page. I also added a `FlashMessage` component and corresponding HOC, which handles the following:
- `connectFlashMessage` connects to the `flash` reducer and renders `FlashMessage` only if there is a flash message defined (using the `key` property)
- `FlashMessage` renders the message in an `Alert` component, and also calls the `clearFlashMessage` action on unmount - this should guarantee flash messages are only shown to a user once, and will be cleared when the user navigates to another page.

I also replaced existing "flash" messages with this new component & reducer:
- Error if service member creation fails
- Error if unable to load user data (but user is logged in)
- Success when submitting a move
  - There was some existing logic to show a `PpmAlert` component here, but I discussed with the design team and we agreed to change to a more generic success message (which is [line 101 in the content doc](https://docs.google.com/spreadsheets/d/14_qlyhSN6-VjXmBLRtBHj2OUwy76NONOEUbYiRYQ2-w/edit#gid=1041135497))

## Reviewer Notes

Things I didn't do:
- Ideally, the `ConnectedFlashMessage` component is rendered at a more global level, such as MyMove/index.jsx (so it renders on every page, not just home). However right now there's some weird organization with the SignIn component for example, and I think we should clean up the global layout markup for the customer app generally, so I wanted to leave that for later.

To test:
- You can easily test the success flash message by submitting a move (with any shipment types) and verifying you see the success message on the home screen. It should then disappear once you click to another page.
- For the error messages, I'm not sure how to cause one of these errors in a real-life scenario. However, you can force the service member creation error if you log in as an existing service member, and then change [line 65 in `src/sagas/onboarding.js`](https://github.com/transcom/mymove/pull/5393/files#diff-6ceafd6b7d1973f6b2c2a2765bc9bdaf5dc05db40b6049076de7aaf04670fdd2L65) to `if (true)` ...then a server error will occur because it's attempting to create a service member for a user that already has one.

## Setup

```
make server_run
make client_run
```

## Code Review Verification Steps

 [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-2978) for this change

## Screenshots

![image](https://user-images.githubusercontent.com/2723066/101062369-30acf400-3557-11eb-8d8f-d59bcbfc28bf.png)

![image](https://user-images.githubusercontent.com/2723066/101061374-0c044c80-3556-11eb-96ea-848c25d4212b.png)
